### ISSUE TITLE
SSO sync throws NoSuchMethodError

### ISSUE THREAD

**Issue by dmatej** [payara/Payara#200]
Oracle JDK8u31, 64bit Kubuntu 14.10, current Payara snapshot (payara-ml.zip, built from commit dmatej:Payara-issue-154) , cluster with two instances and SSO enabled. After successful login on one instance, session synchronization fails.

```
[2015-02-06T23:39:27.212+0100] [Payara 4.1] [SEVERE] [] [javax.enterprise.resource.corba.org.glassfish.enterprise.iiop.impl] [tid: _ThreadID=54 _ThreadName=GMS-processNotify-Group-cis-web-cluster-thread-2] [timeMillis: 1423262367212] [levelValue: 1000] [[
  org.glassfish.grizzly.Connection.write(Ljava/lang/Object;Ljava/lang/Object;Lorg/glassfish/grizzly/CompletionHandler;)Lorg/glassfish/grizzly/GrizzlyFuture;
java.lang.NoSuchMethodError: org.glassfish.grizzly.Connection.write(Ljava/lang/Object;Ljava/lang/Object;Lorg/glassfish/grizzly/CompletionHandler;)Lorg/glassfish/grizzly/GrizzlyFuture;
        at com.sun.enterprise.mgmt.transport.grizzly.grizzly2.GrizzlyTCPMessageSender.send(GrizzlyTCPMessageSender.java:140)
        at com.sun.enterprise.mgmt.transport.grizzly.grizzly2.GrizzlyTCPMessageSender.doSend(GrizzlyTCPMessageSender.java:99)
        at com.sun.enterprise.mgmt.transport.AbstractMessageSender.send(AbstractMessageSender.java:74)
        at com.sun.enterprise.mgmt.transport.grizzly.GrizzlyNetworkManager.send(GrizzlyNetworkManager.java:288)
        at com.sun.enterprise.mgmt.ClusterManager.send(ClusterManager.java:395)
        at com.sun.enterprise.mgmt.ClusterManager.send(ClusterManager.java:419)
        at com.sun.enterprise.ee.cms.impl.base.GroupCommunicationProviderImpl.sendMessage(GroupCommunicationProviderImpl.java:353)
        at com.sun.enterprise.ee.cms.impl.base.DistributedStateCacheImpl.sendMessage(DistributedStateCacheImpl.java:601)
        at com.sun.enterprise.ee.cms.impl.base.DistributedStateCacheImpl.getFromCacheForPattern(DistributedStateCacheImpl.java:371)
        at com.sun.enterprise.ee.cms.impl.common.JoinedAndReadyNotificationSignalImpl.getMemberDetails(JoinedAndReadyNotificationSignalImpl.java:169)
        at org.glassfish.enterprise.iiop.impl.IiopFolbGmsClient.handleSignal(IiopFolbGmsClient.java:239)
        at org.glassfish.enterprise.iiop.impl.IiopFolbGmsClient.processNotification(IiopFolbGmsClient.java:215)
        at com.sun.enterprise.ee.cms.impl.client.JoinedAndReadyNotificationActionImpl.consumeSignal(JoinedAndReadyNotificationActionImpl.java:69)
        at com.sun.enterprise.ee.cms.impl.common.Router$CallableAction.call(Router.java:687)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
```
```
[2015-02-06T23:41:54.943+0100] [Payara 4.1] [WARNING] [] [ShoalLogger] [tid: _ThreadID=329 _ThreadName=pool-10-thread-1] [timeMillis: 1423262514943] [levelValue: 900] [[
  Error during groupHandle.sendMessage(cis-web2, SSOStore; size=619) Cause:org.glassfish.grizzly.Connection.write(Ljava/lang/Object;Ljava/lang/Object;Lorg/glassfish/grizzly/CompletionHandler;)Lorg/glassfish/grizzly/GrizzlyFuture;]]
```

**Comment by dmatej** [comment]
This is most evil annotation in the world today: @SuppressWarnings("unchecked"). And Shoal is full of them, even on classes :(

**Comment by smillidge** [comment]
We need to try and reproduce. Any test case you have would be great. 

Longer term we want to remove Shoal as we don;t think it is being maintained.

https://github.com/payara/Payara/issues/27

**Comment by dmatej** [comment]
Test case (unit test) is a problem, I don't know how to test clusters ...
The cause is in incompatible API of new grizzly-framework and old Shoal. 
How to reproduce:
1) Create ear with several war files using same realm.
2) Set the HTTP Service to use SSO
3) Create cluster with two instances
4) Deploy EAR to the cluster
5) Login to the application in the first instance
6) Access to the application in the second instance
7) You are logged out and there will be this stacktrace in one of server.log files.

Now I have fixed versions of shoal and payara (no changes in grizzly), can you check it?
https://github.com/dmatej/Shoal/tree/payaraFix
https://github.com/dmatej/Payara/tree/Payara-issue-200
(I hope I did not break something else)

**Comment by dmatej** [comment]
Payara with the fixed Shoal works even with really huge application, but Hazelcast sometimes "spams" server.log with this stacktrace (maybe it should be for another issue, I don't know if it can be optimized?). It seems it does not cause any error in the application:
http://docs.hazelcast.org/docs/3.4/javadoc/com/hazelcast/cache/CacheNotExistsException.html
```
[2015-02-14T00:48:33.847+0100] [Payara 4.1] [SEVERE] [] [com.hazelcast.cache.impl.operation.CacheContainsKeyOperation] [tid: _ThreadID=47 _ThreadName=hz.glassfish-web.cis-web2.partition-operation.thread-1] [timeMillis: 1423871313847] [levelValue: 1000] [[
  [192.26.1.2]:5903 [dev] [3.4] Cache is already destroyed or not created yet, on Member [192.26.1.2]:5903 this
com.hazelcast.cache.CacheNotExistsException: Cache is already destroyed or not created yet, on Member [192.26.1.2]:5903 this
        at com.hazelcast.cache.impl.AbstractCacheRecordStore.<init>(AbstractCacheRecordStore.java:106)
        at com.hazelcast.cache.impl.CacheRecordStore.<init>(CacheRecordStore.java:61)
        at com.hazelcast.cache.impl.CacheService.createNewRecordStore(CacheService.java:51)
        at com.hazelcast.cache.impl.CachePartitionSegment$1.createNew(CachePartitionSegment.java:56)
        at com.hazelcast.cache.impl.CachePartitionSegment$1.createNew(CachePartitionSegment.java:53)
        at com.hazelcast.util.ConcurrencyUtil.getOrPutSynchronized(ConcurrencyUtil.java:40)
        at com.hazelcast.cache.impl.CachePartitionSegment.getOrCreateCache(CachePartitionSegment.java:76)
        at com.hazelcast.cache.impl.AbstractCacheService.getOrCreateCache(AbstractCacheService.java:104)
        at com.hazelcast.cache.impl.operation.AbstractCacheOperation.beforeRun(AbstractCacheOperation.java:62)
        at com.hazelcast.spi.impl.BasicOperationService$OperationHandler.handle(BasicOperationService.java:727)
        at com.hazelcast.spi.impl.BasicOperationService$OperationHandler.access$500(BasicOperationService.java:709)
        at com.hazelcast.spi.impl.BasicOperationService$OperationPacketHandler.handle(BasicOperationService.java:683)
        at com.hazelcast.spi.impl.BasicOperationService$OperationPacketHandler.handle(BasicOperationService.java:627)
        at com.hazelcast.spi.impl.BasicOperationService$OperationPacketHandler.access$1500(BasicOperationService.java:614)
        at com.hazelcast.spi.impl.BasicOperationService$BasicDispatcherImpl.dispatch(BasicOperationService.java:566)
        at com.hazelcast.spi.impl.BasicOperationScheduler$OperationThread.process(BasicOperationScheduler.java:466)
        at com.hazelcast.spi.impl.BasicOperationScheduler$OperationThread.doRun(BasicOperationScheduler.java:458)
        at com.hazelcast.spi.impl.BasicOperationScheduler$OperationThread.run(BasicOperationScheduler.java:432)
]]
```

**Comment by smillidge** [comment]
Great work, I'll test the patched shoal on web session replication. What are you using Hazelcast for in your application?

**Comment by dmatej** [comment]
Only for simple caching, application only creates some named caches and put and return items. It seems the exception is thrown only once for each cache but everything works.
